course/api/full_serializers.py

```diff
@@ -105,6 +109,101 @@ class Meta(CourseBriefSerializer.Meta):
         )
 
 
+class CourseWriteSerializer(AplusModelSerializer):
+    """
+    Serializer for creating and modifying course models using POST and PUT.
+    """
+    code = serializers.CharField()
+    name = serializers.CharField()
+    course_url = serializers.CharField()
+    teachers = serializers.ListSerializer(child=serializers.CharField())
+
+    class Meta(AplusModelSerializer.Meta):
+        model = CourseInstance
+        fields = (
+            'code',
+            'name',
+            'course_url',
+            'url',
+            'instance_name',
+            'language',
+            'starting_time',
+            'ending_time',
+            'visible_to_students',
+            'configure_url',
+            'teachers',
+        )
+
+    def set_teachers(self, course: CourseInstance, teachers: list) -> None:
+        users = []
+        for i in teachers:
+            userprofile: UserProfile = None
+            try:
+                user = User.objects.get(username=i)
+            except User.DoesNotExist:
+                # If user does not exist, create a new user.
+                # If external authentication (e.g. Shibboleth) is used, other
+                # attributes will be updated when user logs in for the first time.
+                user = User.objects.create_user(i)
+
+            userprofile = UserProfile.objects.get(user=user)
```

Comment:
You don't need to make an explicit query to model UserProfile since Django automatically reverses the OneToOne relationship: `userprofile = user.userprofile`